chain-signatures/node/src/indexer_eth.rs

```diff
@@ -368,29 +416,79 @@ pub async fn run(
         .await
         .map_err(|err| anyhow::anyhow!("Failed to subscribe to new block heads: {:?}", err))?;
 
-    let mut interval = interval(Duration::from_millis(100));
-    let mut blocks_failed: VecDeque<u64> = VecDeque::new();
-    const MAX_BLOCK_RETRIES_PER_TICK: usize = 10;
-    loop {
-        for _ in 0..blocks_failed.len().min(MAX_BLOCK_RETRIES_PER_TICK) {
-            if let Some(block_number) = blocks_failed.pop_front() {
-                if let Err(err) = process_block(
-                    block_number,
-                    &client,
-                    eth_contract_addr,
-                    sign_tx.clone(),
-                    node_near_account_id.clone(),
-                )
-                .await
-                {
-                    tracing::warn!("Retry failed for block {block_number}: {:?}", err);
-                    blocks_failed.push_back(block_number);
-                } else {
-                    tracing::info!("Successfully retried block: {}", block_number);
-                }
-            }
+    let blocks_failed: Arc<RwLock<VecDeque<(u64, alloy::primitives::B256)>>> =
+        Arc::new(RwLock::new(VecDeque::new()));
+
+    let requests_indexed: Arc<RwLock<VecDeque<BlockAndRequests>>> =
+        Arc::new(RwLock::new(VecDeque::new()));
+
+    let finalized_epoch: Arc<RwLock<HashMap<u64, alloy::primitives::B256>>> =
+        Arc::new(RwLock::new(HashMap::new()));
+    let finalized_block: Arc<RwLock<u64>> = Arc::new(RwLock::new(0));
+
+    let client = Arc::new(client);
+    let client_clone = Arc::clone(&client);
+    let finalized_block_clone = Arc::clone(&finalized_block);
+    let finalized_epoch_clone = Arc::clone(&finalized_epoch);
+    tokio::spawn(async move {
+        tracing::info!("Spawned refresh task started");
+        loop {
+            refresh_finalized_epoch(
+                &client_clone,
+                &untrusted_rpc_client,
+                &finalized_block_clone,
+                &finalized_epoch_clone,
+            )
+            .await
+            .unwrap_or_else(|err| {
+                tracing::warn!("Failed to refresh finalized epoch: {:?}", err);
+            });
+            tokio::time::sleep(Duration::from_millis(eth.refresh_finalized_interval)).await;
```

Comment:
We started using `tick` pattern in our code base, let's stick to it here and in other places.